### ISSUE TITLE
Strange results

### ISSUE THREAD

**Issue by kaykyr** [jinglescode/time-series-forecasting-tensorflowjs#7]
Hello this is a very interesting project. I tried with the default graphics (MSFT) and I got a great result.

But now, I am trying to train a model for EURUSD asset from Forex (15 minutes timeframe)

![Screen Shot 2021-04-08 at 23 58 33](https://user-images.githubusercontent.com/30188217/114122379-865b2980-98c6-11eb-9be6-b830f0e3c391.png)

But I am getting this result... and sometimes is just a rect line...

I am trying to understand this better, I am not experienced with neural networks before, but may you able to help me to understand?

Thank you!

**Comment by kaykyr** [comment]
Or may someone can recommend any content to study and understand it better? Thanks

**Comment by jinglescode** [comment]
Hi, could you upload a more complete picture of the graph, with the legends and the axis. And also the parameters you have used.

**Comment by kaykyr** [comment]
> Hi, could you upload a more complete picture of the graph, with the legends and the axis. And also the parameters you have used.

Hello! In attach the JSON containing EURUSD (~6.000 5 minutes candles) data:
https://drive.google.com/file/d/1XXiUfSL9eBkPjUhAnu4mTkH1jdceieTz/view?usp=sharing

I am using the same algorithm with the following parameters:

![Screen Shot 2021-04-09 at 00 40 09](https://user-images.githubusercontent.com/30188217/114125102-2a939f00-98cc-11eb-8066-fb38b4c4e387.png)

Training Dataset Size: 98
Epochs: 25
Learning Rate: 0,01
Hidden LSTM Layers: 4
Simple Moving Average Period: 20

![Screen Shot 2021-04-09 at 00 40 37](https://user-images.githubusercontent.com/30188217/114125146-4139f600-98cc-11eb-91f2-fe798c36386c.png)


I also tried with different training dataset size, epochs and SMA periods... The result are always the same.

This is the training chart:
![Screen Shot 2021-04-09 at 00 46 52](https://user-images.githubusercontent.com/30188217/114125556-18fec700-98cd-11eb-86b3-ccbd49190c6e.png)

This is the validation chart:
![Screen Shot 2021-04-09 at 00 47 23](https://user-images.githubusercontent.com/30188217/114125591-2caa2d80-98cd-11eb-9dc5-154b4908f63a.png)

And this is the predict chart:
![Screen Shot 2021-04-09 at 00 47 51](https://user-images.githubusercontent.com/30188217/114125618-3b90e000-98cd-11eb-8d87-f6507df293d7.png)

This trained model download:
https://drive.google.com/file/d/1n7eMn-OTMK1ziSqI-2IOKOFR3jLrm_ZH/view?usp=sharing

**Comment by fgabrielbr** [comment]
I am facing exactly the same problem

**Comment by clarklight** [comment]
I am seeing the same result as well, i forked a version and trained the code, i see the same result locally as well. 
The only difference between your initial commit is this 2 lines. 

  let X = inputs.slice(0, Math.floor(trainingsize / 100 * inputs.length));
  let Y = outputs.slice(0, Math.floor(trainingsize / 100 * outputs.length));

 The current version doesnt have that, not sure if its anything to do with the issue. (I am still learning, so not sure if thats the issue). 

Also on this link https://jinglescode.github.io/time-series-forecasting-tensorflowjs/
Followed all the settings you used on your readme, and i also got a flatline for validating the prediction.

**Comment by jinglescode** [comment]
This is the challenge of machine learning, that a model, out of the box, doesn't work for all kinds of data. There is a lot of parameter tuning involve. There may need to do more feature engineering. Or there is a need to change the model architecture. 

Indeed, you can see that you are not the only one who faced this problem:

- [LSTM always predict same values for all inputs in the batch](https://discuss.pytorch.org/t/lstm-always-predict-same-values-for-all-inputs-in-the-batch/53082)
- [Outputs of LSTM Model all Very Similar or the Same](https://discuss.pytorch.org/t/outputs-of-lstm-model-all-very-similar-or-the-same/84370)
- [Pytorch RNN always gives the same output for multivariate time series](https://www.reddit.com/r/learnmachinelearning/comments/9zugw8/pytorch_rnn_always_gives_the_same_output_for/)
- [PyTorch regression is producing the same numbers as prediction](https://stackoverflow.com/questions/59506405/pytorch-regression-is-producing-the-same-numbers-as-prediction)

I do not have an answer for you, unfortunately. Some suggestions would be:
- try tweaking the normalize the data portion (`.div(tf.scalar(10)` might be causing problems [line 19 and 20](https://github.com/jinglescode/time-series-forecasting-tensorflowjs/blob/master/model.js))
- include more data features, maybe can include volume as well
- include more data features, maybe trading indicators such as MACD, RSI, Bollinger bands
- build another model architecture

**Comment by jinglescode** [comment]
@kaykyr, it's really strange you are getting a flat line on the training set too. I think have to debug (doing lots of `console.log`) and figure out what is the issue. At the minimum, you should get the model to overfit on the training set.

**Comment by kaykyr** [comment]
> @kaykyr, it's really strange you are getting a flat line on the training set too. I think have to debug (doing lots of `console.log`) and figure out what is the issue. At the minimum, you should get the model to overfit on the training set.

Hey, thank you!

It's very strange, but I built the same model with python with latest versions (Keras/Tensorflow) I got other result, but looks like it's predicting to the future the same thing of the past.

But, it's ok, now I know that this is a thing that needs too much effort to works great, I'll keep learning and improving my model.
